lib/Target/ARM/ARMPagerandoOptimizer.cpp

```diff
@@ -0,0 +1,265 @@
+// This pass optimizes calls inside the same position-independent bin to direct
+// calls to avoid the overhead of indirect calls through the POT.
+//
+//===----------------------------------------------------------------------===//
+
+#include "ARM.h"
+#include "ARMBaseInstrInfo.h"
+#include "ARMBaseRegisterInfo.h"
+#include "ARMConstantPoolValue.h"
+#include "ARMMachineFunctionInfo.h"
+#include "llvm/CodeGen/MachineRegisterInfo.h"
+#include "llvm/Target/TargetLowering.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "pagerando"
+
+namespace {
+class PagerandoOptimizer : public MachineFunctionPass {
+public:
+  static char ID;
+  explicit PagerandoOptimizer() : MachineFunctionPass(ID) {}
+
+  bool runOnMachineFunction(MachineFunction &MF) override;
+
+  MachineFunctionProperties getRequiredProperties() const override {
+    return MachineFunctionProperties().set(
+        MachineFunctionProperties::Property::TracksLiveness);
+  }
+
+private:
+  void optimizeCall(MachineInstr *MI, const Function *Callee);
+  void replaceWithDirectCall(MachineInstr *MI, const Function *Callee);
+  void replaceWithPCRelativeCall(MachineInstr *MI, const Function *Callee);
+  void deleteCPEntries(MachineFunction &MF, const SmallSet<int, 8> &Workset);
+};
+} // end anonymous namespace
+
+char PagerandoOptimizer::ID = 0;
+INITIALIZE_PASS(PagerandoOptimizer, "pagerando-optimizer",
+                "Pagerando intra-bin optimizer", false, false)
+
+FunctionPass *llvm::createPagerandoOptimizerPass() {
+  return new PagerandoOptimizer();
+}
+
+static bool isIntraBin(const MachineConstantPoolEntry &E, StringRef BinPrefix) {
+  if (!E.isMachineConstantPoolEntry()) return false;
+
+  // ARMConstantPoolValue lacks casting infrastructure to use dyn_cast directly
+  auto *CPV = static_cast<ARMConstantPoolValue*>(E.Val.MachineCPVal);
+  auto *CPC = dyn_cast<ARMConstantPoolConstant>(CPV);
+  if (!CPC) return false;
+
+  auto M = CPC->getModifier();
+  auto *F = dyn_cast_or_null<Function>(CPC->getGV());
+
+  return (M == ARMCP::POTOFF || M == ARMCP::BINOFF)
+      && F &&F->getSectionPrefix() == BinPrefix;
+}
+
+static const Function *getCallee(const MachineConstantPoolEntry &E) {
+  auto *CPC = static_cast<ARMConstantPoolConstant*>(E.Val.MachineCPVal);
+  return cast<Function>(CPC->getGV());
+}
+
+static int getCPIndex(const MachineInstr &MI) {
+  if (MI.mayLoad() && MI.getNumOperands() > 1 && MI.getOperand(1).isCPI()) {
+    return MI.getOperand(1).getIndex();
+  }
+  return -1;
+}
+
+bool PagerandoOptimizer::runOnMachineFunction(MachineFunction &MF) {
+  // This pass is an optimization (optional), therefore check skipFunction.
+  if (skipFunction(*MF.getFunction()) || !MF.getFunction()->isPagerando()) {
+    return false;
+  }
+
+  auto BinPrefix = MF.getFunction()->getSectionPrefix().getValue();
+  auto &CPEntries = MF.getConstantPool()->getConstants();
+
+  // Find intra-bin CP entries
+  SmallSet<int, 8> Workset;
+  for (int Index = 0; Index < CPEntries.size(); ++Index) {
```

Comment:
LLVM style says to compute the end value in the loop initializer: https://github.com/rinon/icfp2017/blob/master/src/punter.rs#L156